README.md
# Kontist SDK

JavaScript SDK for connecting to Kontist using OAuth2 and GraphQL. Please see https://kontist.dev/sdk for a more complete documentation.

## Installation

Add as dependency to your project:

```bash
npm install kontist
```

You will need a valid client id and setup your redirect uri for authentication. You may request your client id in the API Client Management on https://kontist.dev/client-management/.

## Usage (NodeJS / TypeScript)

```typescript
import express from "express";
import { Client } from "kontist";

const CALLBACK_PATH = "/auth/callback";
const REDIRECT_URI = <YOUR_BASE_URL> + CALLBACK_PATH;
const clientSecret = <YOUR_CLIENT_SECRET>;
const state = (Math.random() + "").substring(2);
const app = express();

// create a client
const client = new Client({
  clientId: "YOUR_CLIENT_ID",
  redirectUri: REDIRECT_URI,
  scopes: ["transactions"],
  clientSecret,
  state
});

// redirect not authenticated user to Kontist form
app.get("/auth", async (req, res) => {
  const uri = await client.auth.tokenManager.getAuthUri();
  res.redirect(uri);
});

// get user token data
app.get(CALLBACK_PATH, async (req, res) => {
  const callbackUrl = req.originalUrl;

  try {
    const token = await client.auth.tokenManager.fetchToken(callbackUrl);
    /* got access token, login successful */
    res.send("Successful, your token is " + token.accessToken);
  } catch (e) {
    /* handle error */
    res.send("Failed: " + JSON.stringify(e));
  }
});

app.listen(3000, function() {
  console.log("Listening on port 3000!");
});
```

You should be able to issue new accessToken by simply calling:

```typescript
const token = await client.auth.tokenManager.refresh();
```

Optionally, this method accepts a number as an argument to specify after how many milliseconds the refresh request should timeout (default is 10000):

```typescript
// abort after 20 seconds
const token = await client.auth.tokenManager.refresh(20000);
```

**Note:** to create a client that can handle public request, you need to initialize `Client` without `clientId`.

```typescript
const client = new Client({
  scopes: ["transactions"],
  state
});
```

## Usage (Browser)

```html
<html>
  <body>
    <script src="https://cdn.kontist.com/sdk.min.js"></script>
    <script>
      // persist a random value
      sessionStorage.setItem(
        "state",
        sessionStorage.getItem("state") || (Math.random() + "").substring(2)
      );
      sessionStorage.setItem(
        "verifier",
        sessionStorage.getItem("verifier") || (Math.random() + "").substring(2)
      );

      // initialize Kontist client
      const client = new Kontist.Client({
        clientId: "<your client id>",
        redirectUri: "<your base url>",
        scopes: ["transactions"],
        state: sessionStorage.getItem("state"),
        verifier: sessionStorage.getItem("verifier")
      });

      const params = new URL(document.location).searchParams;
      const code = params.get("code");
      if (!code) {
        // page not called with "code" query parameter, let's redirect the user to the login
        client.auth.tokenManager.getAuthUri().then(function(url) {
          window.location = url;
        });
      } else {
        // we have a code, the client now can fetch a token
        client.auth.tokenManager.fetchToken(document.location.href).then(function() {
          // do a simple graphql query and output the account id
          client.graphQL
            .rawQuery(
              `{
              viewer {
                mainAccount {
                  iban
                  balance
                }
              }
            }`
            )
            .then(function(result) {
              console.log(result);
            });
        });
      }
    </script>
  </body>
</html>
```

Kontist SDK allows renewing access tokens in browser environments using this simple method:

```typescript
const token = await client.auth.tokenManager.refresh();
```

Optionally, this method accepts a number as an argument to specify after how many milliseconds the refresh request should timeout (default is 10000):

```typescript
// abort after 20 seconds
const token = await client.auth.tokenManager.refresh(20000);
```

### GraphQL queries

#### Raw

```typescript
const query = `{
  viewer {
    mainAccount {
      id
    }
  }
}`;

const result = await client.graphQL.rawQuery(query);
```

#### Transactions

An example to show how to fetch all user transactions:

```typescript
let transactions = [];
for await (const transaction of client.models.transaction.fetchAll()) {
  transactions = transactions.concat(transaction);
}
```

To fetch up to 50 latest transactions:

```typescript
const transactions = await client.models.transaction.fetch();
```

To subscribe to new transactions:

```typescript
const onNext = transaction => {
  // do something with the transaction
}

const onError = error => {
  // do something with the error
}

client.models.transaction.subscribe(onNext, onError);
```

Whenever a new transaction is created, the `onNext` function will be called.
Whenever an error occurs with the subscription, the `onError` function will be called.

The `subscribe` method returns a `Subscription` object with an `unsubscribe` method to be called when you want to unsubscribe to new transactions:

```typescript
const { unsubscribe } = client.models.transaction.subscribe(onNext, onError);
// ...
unsubscribe();
// after this point, onNext / onError will no longer be called when new transactions / errors are received
```

#### Transfers

To create and confirm a transfer / timed order / standing order:

```typescript
const confirmationId = await client.models.transfer.createOne({
  amount: <amount>,
  recipient: <recipent_name>,
  iban: <recipent_iban>,
  purpose: <optional_description>,
  e2eId: <optional_e2eId>,
  executeAt: <optional_order_execution_date> // mandatory for timed and standing order
  lastExecutionDate: <optional_last_execution_date> // optional for standing order
  reoccurrence: <optional_order_reoccurrence> // mandatory for standing order
});

// wait for sms
const smsToken = ...

const result = await client.models.transfer.confirmOne(
  confirmationId,
  smsToken
);
```

To create and confirm multiple transfers (with only one confirmation):

```typescript
const confirmationId = await client.models.transfer.createMany([{
  amount: <amount>,
  recipient: <recipent_name>,
  iban: <recipent_iban>,
  purpose: <optional_description>,
  e2eId: <optional_e2eId>,
}, {
  amount: <amount>,
  recipient: <recipent_name>,
  iban: <recipent_iban>,
  purpose: <optional_description>,
  e2eId: <optional_e2eId>,
}]);

// wait for sms
const smsToken = ...

const result = await client.models.transfer.confirmMany(
  confirmationId,
  smsToken
);
```

An example to show how to fetch all user transfers of a given type:

```typescript
let transfers = [];
for await (const transfer of client.models.transfer.fetchAll({
  type: Schema.TransferType.SepaTransfer
})) {
  transfers = transfers.concat(transfer);
}
```

To fetch up to 50 latest transfers of a given type:

```typescript
const transfers = await client.models.transfer.fetch({
  type: Schema.TransferType.TimedOrder
});
```

#### Cards

Several methods are available to create, manage and fetch cards.

Here is an example of how to create a new card for a user:

```typescript
const card = await client.models.card.create({
  type: CardType.VisaBusinessDebit
});
```

And here is how you can retrieve the details of existing cards:

```typescript
const result = await client.models.card.fetch();
```

## MFA (Multi-Factor Authentication)

Accessing Kontist banking APIs require Multi-Factor Authentication (MFA).

MFA is available once you have installed the Kontist application and paired your device in it.

The following steps are necessary to complete the MFA procedure:
1. initiate the procedure by creating a challenge (Kontist SDK exposes a method to do that)
2. click the push notification you received on your phone, it will open the Kontist application
3. login (if applicable) and confirm the MFA by clicking on the corresponding button

Kontist SDK exposes a method to initiate the push notification MFA flow after you successfully received the initial access token:

```typescript
// fetch a regular access token
const token = await client.auth.tokenManager.fetchToken(callbackUrl);

try {
  // create a push notification challenge and wait for confirmation
  const confirmedToken = await client.auth.push.getConfirmedToken();
  // once it has been verified, your `client` instance will have a confirmed access token
  // the confirmed token is also returned in case you want to store it
} catch (err) {
  // if the challenge expires, a `ChallengeExpiredError` will be thrown
  // if the challenge is denied, a `ChallengeDeniedError` will be thrown
  console.log(err);
}
```

After obtaining a confirmed auth token with this method, you will have access to all banking APIs.

If you want to cancel a pending push notification confirmation, you can call the following method:

```typescript
client.auth.push.cancelConfirmation();
```

The Promise returned by `getConfirmedToken` will then reject with a `MFAConfirmationCanceledError`.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
## [0.50.4] - 2023-09-20
## Revert remove
- Put back `title, description, button, featuresToggleLabel, featureGroups` from the `subscriptionPlans` query, for backward compatibility
- Update the schema with the above deprecated fields

## [0.50.3] - 2023-09-12
## Removed
- Removed `title, description, button, featuresToggleLabel, featureGroups` from the `subscriptionPlans` query

## [0.49.0] - 2023-01-31
## Added
- Updated types in schema
- Added `getPinKey` and `changePINEncrypted` methods to `Card` module

## [0.48.0] - 2022-09-16
## Added
- Updated types in schema
- Added solarisBalance graphQL API
## [0.46.2] - 2022-07-05
- Updated types in schema
## [0.46.0] - 2022-06-08
### Added
- Added `id` field to `category` in document metadata query

## [0.46.0] - 2022-06-08
### Added
- Added `metadata` field to `Document` type

## [0.45.0] - 2022-05-05
### Added
- Added `EmailDocument` class with `fetchOne`, `fetchAll` and `matchEmailDocumentToTransaction` functions
## [0.44.0] - 2022-04-06
### Added
- Added `hasBusinessTaxNumber` field to `User` type
- Added `hasBusinessTaxNumberUpdatedAt` field to `User` type
## [0.43.0] - 2022-03-24

- Downgrades `client-oauth2` to `4.3.0` for compatibility reasons
## [0.38.0] - 2022-03-15
### Added
- Added `discountAmount` field to `Money` type
- Added `couponValidFor` field to `SubscriptionPlansResponse` type
## [0.36.0] - 2021-08-13
### Changed
- Replace `elsterCode` by `categoryCode`. Keep `elsterCode` for backward compatibility.

## [0.35.41] - 2021-06-29
- Remove deprecated `categorizeTransaction` mutation
### Added
- Added `isYearlyTaxServiceOnboardingCompleted` field to `User` type
## [0.35.26] - 2021-03-24
### Added
- Added `isYearlyTaxServiceOnboardingCompleted` field to `User` type

## [0.35.23] - 2021-03-15
### Added
- Added `invoicePdf` field to `User` type

## [0.35.20] - 2021-03-05
### Added
- Added `createInvoiceLogo` and `deleteInvoiceLogo` mutations to GraphQL schema
- Added `logoUrl` field to `InvoiceSettingsOutput` type

## [0.35.16] - 2021-03-01
### Changed
- Allow making GQL request to public endpoint [#153](https://github.com/kontist/js-sdk/pull/153)

### Added
- Added `poaUrl` field in schema for `viewer` [#169](https://github.com/kontist/js-sdk/pull/169)
- Added `signPOA` mutation to schema [#169](https://github.com/kontist/js-sdk/pull/169)

## [0.35.15] - 2021-02-24
### Added
- Added `categorizationType` field of `TransactionSplit` to the schema and the transactions query [#167](https://github.com/kontist/js-sdk/pull/167)

## [0.34.6] - 2020-09-02
### Changed
- Build correct verifier during oauth handshake [#145](https://github.com/kontist/js-sdk/pull/145)

## [0.33.0] - 2020-08-18

### Changed
- *Breaking*: removed several transaction fields from the results of the fetch and fetchAll transaction methods (see `TRANSACTION_DETAILS` in `lib/graphql/transaction` for the details) [#132](https://github.com/kontist/js-sdk/pull/132)

### Added
- Added `fetchOne` method on transaction model which will return all transaction fields for a given transaction ID [#132](https://github.com/kontist/js-sdk/pull/132)

## [0.31.3] - 2020-07-16

### Added
- Added `isSelfEmployed` field in schema for `UserUpdateInput` and `viewer`

## [0.31.2] - 2020-07-10

### Changed
- Move graphql back to dependencies [#122](https://github.com/kontist/js-sdk/pull/122)

## [0.31.0] - 2020-07-10

### Changed
- Updated dependencies (client-oauth2, graphql, graphql-request, ws) and dev dependencies
- Added cross-fetch ponyfill
- Resolve schema issues with transaction assets

## [0.30.0] - 2020-07-08

### Added
- Added helpers for uploading and deleting transaction assets (`models.transaction.createTransactionAsset`, `models.transaction.finalizeTransactionAssetUpload`, `models.transaction.deleteTransactionAsset`)

### Changed
- `updatePlan` method of `subscription` model extended with optional `couponCode` param

## [0.29.1] - 2020-06-11

### Added
- Added a new `personalNote` field on transactions and transfers - this field allows users to annotate their transactions with text that will only be available to them
- Added a new `update` method on `transaction` model, which allows to update category, userSelectedBookingDate and personalNote for a given transaction

### Changed
- `categorize` method of `transaction` model is now deprecated in favor of `update`

## [0.28.7] - 2020-06-09

### Added
- Added `createdAt` field to transactions

### Changed
- *Breaking*: renamed `taxPastYearAmount` field to `taxPastYearsAmount` for `getStats` method of `account` model

## [0.28.6] - 2020-05-29

### Changed
- *Breaking*: removed `taxCutoffLine` field from `user` model `get` method

## [0.28.3] - 2020-05-06

### Changed
- Extended Overdraft with requestedAt field

## [0.28.2] - 2020-05-05

### Added
- Added mutations for adding/deleting card token reference ids for Google Pay

### Changed
- card query now includes an array of added Google Pay card tokens

## [0.28.1] - 2020-05-04

### Changed
- Update vulnerable dependencies (`yargs`, `yargs-parser`)

## [0.28.0] - 2020-05-04

### Added
- Add helpers to split transactions (`models.transaction.createSplit`, `models.transaction.deleteSplit`, `models.transaction.updateSplit`)
- Add method to fetch account stats (`models.account.getStats`)

### Changed
- *Breaking*: renamed `StandingOrderReoccurenceType` to `StandingOrderReoccurrenceType`

## [0.27.0] - 2020-04-20
### Added
- Add flow type definitions in `lib/graphql/schema.flow.js`

## [0.26.0] - 2020-04-07
### Added
- Add `subscription` model with a method to update subscription plan (`subscription.updatePlan`)

## [0.25.8] - 2020-03-10
### Changed

lib/auth.ts
import * as ClientOAuth2 from "client-oauth2";
import { sha256 } from "js-sha256";
import { btoa } from "abab";
import {
  ClientOpts,
  Challenge,
  ChallengeStatus,
  HttpMethod,
  CreateDeviceParams,
  CreateDeviceResult,
  VerifyDeviceParams,
  DeviceChallenge,
  VerifyDeviceChallengeParams,
  VerifyDeviceChallengeResult,
  GetAuthUriOpts,
  TimeoutID
} from "./types";
import { authorizeSilently } from "./utils";
import {
  ChallengeExpiredError,
  ChallengeDeniedError,
  MFAConfirmationCanceledError,
  UserUnauthorizedError,
  RenewTokenError,
  KontistSDKError
} from "./errors";

import "cross-fetch/polyfill";

export const MFA_CHALLENGE_PATH = "/api/user/mfa/challenges";
export const CREATE_DEVICE_PATH = "/api/user/devices";
export const VERIFY_DEVICE_PATH = (deviceId: string) =>
  `/api/user/devices/${deviceId}/verify`;
export const CREATE_DEVICE_CHALLENGE_PATH = (deviceId: string) =>
  `/api/user/devices/${deviceId}/challenges`;
export const VERIFY_DEVICE_CHALLENGE_PATH = (
  deviceId: string,
  challengeId: string
) => `/api/user/devices/${deviceId}/challenges/${challengeId}/verify`;

const CHALLENGE_POLL_INTERVAL = 3000;
const DEFAULT_TOKEN_REFRESH_TIMEOUT = 10000;

const HTTP_STATUS_NO_CONTENT = 204;

export class Auth {
  private oauth2Client: ClientOAuth2;
  private _token: ClientOAuth2.Token | null = null;
  private baseUrl: string;
  private state?: string;
  private verifier?: string;
  private challengePollInterval: number = CHALLENGE_POLL_INTERVAL;
  private challengePollTimeoutId?: TimeoutID;
  private rejectMFAConfirmation: Function | null = null;

  /**
   * Client OAuth2 module instance.
   *
   * @param baseUrl  Kontist API base url
   * @param opts     OAuth2 client data including at least clientId, redirectUri,
   *                 scopes, state and clientSecret or code verifier (for PKCE).
   * @throws         throws when both clientSecret and code verifier are provided
   */
  constructor(baseUrl: string, opts: ClientOpts) {
    const {
      clientId,
      clientSecret,
      oauthClient,
      redirectUri,
      scopes,
      state,
      verifier
    } = opts;
    this.verifier = verifier;
    this.baseUrl = baseUrl;
    this.state = state;

    if (verifier && clientSecret) {
      throw new KontistSDKError({
        message:
          "You can provide only one parameter from ['verifier', 'clientSecret']."
      });
    }

    if (verifier && (!state || !redirectUri)) {
      throw new KontistSDKError({
        message:
          "If you are providing a 'verifier', you must also provide 'state' and 'redirectUri' options."
      });
    }

    this.oauth2Client =
      oauthClient ||
      new ClientOAuth2({
        accessTokenUri: `${baseUrl}/api/oauth/token`,
        authorizationUri: `${baseUrl}/api/oauth/authorize`,
        clientId,
        clientSecret,
        redirectUri,
        scopes,
        state
      });
  }

  /**
   * Build a uri to which the user must be redirected for login.
   */
  public getAuthUri = async (opts: GetAuthUriOpts = {}): Promise<string> => {
    const query: {
      [key: string]: string | string[];
    } = {
      ...(opts.query || {})
    };

    if (this.verifier) {
      // Implemented according to https://tools.ietf.org/html/rfc7636#appendix-A
      const challenge = (
        btoa(String.fromCharCode.apply(null, sha256.array(this.verifier))) || ""
      )
        .split("=")[0]
        .replace("+", "-")
        .replace("/", "_");

      query.code_challenge = challenge;
      query.code_challenge_method = "S256";
    }

    return this.oauth2Client.code.getUri({ query });
  };

  /**
   * This method must be called during the callback via `redirectUri`.
   *
   * @param callbackUri  `redirectUri` containing OAuth2 data after user authentication
   * @returns            token object which might contain token(s), scope(s), token type and expiration time
   */
  public fetchToken = async (
    callbackUri: string
  ): Promise<ClientOAuth2.Token> => {
    const options: {
      body?: {
        code_verifier: string;
      };
    } = {};

    if (this.verifier) {
      options.body = {
        code_verifier: this.verifier
      };
    }

    const token = await this.oauth2Client.code.getToken(callbackUri, options);

    this._token = token;

    return token;
  };

  /**
   * Fetches token from owner credentials.
   * Only works for client IDs that support the 'password' grant type
   *
   * @param options     Username, password, and an optional set of scopes
   *                    When given a set of scopes, they override the default list of
   *                    scopes of `this` intance
   *
   * @returns           token object which might contain token(s), scope(s), token type and expiration time
   */
  public fetchTokenFromCredentials = async (options: {
    username: string;
    password: string;
    scopes?: string[];
  }) => {
    const getTokenOpts = options.scopes ? { scopes: options.scopes } : {};
    const token = await this.oauth2Client.owner.getToken(options.username, options.password, getTokenOpts);

    this._token = token;

    return token;
  }

  /**
   * Refresh auth token silently for browser environments
   * Renew auth token
   *
   * @param timeout  optional timeout for renewal in ms
   */
  public refresh = async (
    timeout: number = DEFAULT_TOKEN_REFRESH_TIMEOUT
  ): Promise<ClientOAuth2.Token> =>
    this.verifier
      ? this.renewWithWebMessage(timeout)
      : this.renewWithRefreshToken(timeout);

  /**
   * Renew auth token using refresh token
   *
   * @param timeout  timeout for renewal in ms
   */
  private renewWithRefreshToken = async (
    timeout: number
  ): Promise<ClientOAuth2.Token> => {
    return new Promise(async (resolve, reject) => {
      if (!this.token) {
        throw new UserUnauthorizedError();
      }

      const timeoutId = setTimeout(() => {
        reject(
          new RenewTokenError({
            message: "Server did not respond with a new auth token, aborting."
          })
        );
      }, timeout);

      let token;
      try {
        token = await this.token.refresh();
      } catch (error) {
        return reject(
          new RenewTokenError({
            message: error.message
          })
        );
      }

      clearTimeout(timeoutId);

      this._token = token;
      return resolve(token);
    });
  };

  /**
   * Renew auth token for browser environments using web_message response mode and prompt none
   *
   * @param timeout  timeout for renewal in ms
   */
  private renewWithWebMessage = async (
    timeout: number
  ): Promise<ClientOAuth2.Token> => {
    if (!document || !window) {
      throw new RenewTokenError({
        message:
          "Web message token renewal is only available in browser environments"
      });
    }

    const iframeUri = await this.getAuthUri({
      query: {
        prompt: "none",
        response_mode: "web_message"
      }
    });

    try {
      const code = await authorizeSilently(iframeUri, this.baseUrl, timeout);
      const fetchTokenUri = `${
        document.location.origin
      }?code=${code}&state=${encodeURIComponent(this.state || "")}`;
      const token = await this.fetchToken(fetchTokenUri);

      return token;
    } catch (error) {
      throw new RenewTokenError({
        message: error.message
      });
    }
  };

  /**
   * Sets up  previously created token for all upcoming requests.
   *
   * @param accessToken   access token
   * @param refreshToken  optional refresh token
   * @param tokenType     token type
   * @returns             token object which might contain token(s), scope(s), token type and expiration time
   */
  public setToken = (
    accessToken: string,
    refreshToken?: string,
    tokenType?: string
  ): ClientOAuth2.Token => {
    const data = {};
    let token;

    if (tokenType && refreshToken) {
      token = this.oauth2Client.createToken(
        accessToken,
        refreshToken,
        tokenType,
        data
      );
    } else if (refreshToken) {
      token = this.oauth2Client.createToken(accessToken, refreshToken, data);
    } else {
      token = this.oauth2Client.createToken(accessToken, data);
    }

    this._token = token;

    return token;
  };

  /**
   * Perform a request against Kontist REST API
   */
  private request = async (
    path: string,
    method: HttpMethod,
    body?: string | Object
  ) => {
    if (!this.token) {
      throw new UserUnauthorizedError();
    }

    const requestUrl = new URL(path, this.baseUrl).href;

    const response = await fetch(requestUrl, {
      method,
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json",
        Authorization: `Bearer ${this.token.accessToken}`
      },
      body: JSON.stringify(body)
    });

    if (!response.ok) {
      throw new KontistSDKError({
        status: response.status,
        message: response.statusText
      });
    }

    if (response.status === HTTP_STATUS_NO_CONTENT) {
      return;
    }

    return response.json();
  };

  /**
   * Called by `getMFAConfirmedToken`. Calls itself periodically
   * until the challenge expires or its status is updated
   */
  private pollChallengeStatus = (
    pendingChallenge: Challenge,
    resolve: Function,
    reject: Function
  ) => async () => {
    let challenge;
    try {
      challenge = await this.request(
        `${MFA_CHALLENGE_PATH}/${pendingChallenge.id}`,
        HttpMethod.GET
      );
    } catch (error) {
      return reject(error);
    }

    this.rejectMFAConfirmation = null;

    const hasExpired = new Date(challenge.expiresAt) < new Date();
    const wasDenied = challenge.status === ChallengeStatus.DENIED;
    const wasVerified = challenge.status === ChallengeStatus.VERIFIED;

    if (hasExpired) {
      return reject(new ChallengeExpiredError());
    } else if (wasDenied) {
      return reject(new ChallengeDeniedError());
    } else if (wasVerified) {
      const { token: confirmedToken } = await this.request(
        `${MFA_CHALLENGE_PATH}/${challenge.id}/token`,
        HttpMethod.POST
      );

      const token = this.setToken(confirmedToken);
      return resolve(token);
    }

    this.rejectMFAConfirmation = reject;
    this.challengePollTimeoutId = setTimeout(
      this.pollChallengeStatus(pendingChallenge, resolve, reject),
      this.challengePollInterval
    );
  };

  /**
   * Create an MFA challenge and request a confirmed access token when verified
   */
  public getMFAConfirmedToken = async () => {
    const challenge = await this.request(MFA_CHALLENGE_PATH, HttpMethod.POST);

    return new Promise((resolve, reject) =>
      this.pollChallengeStatus(challenge, resolve, reject)()
    );
  };

  /**
   * Clear pending MFA confirmation
   */
  public cancelMFAConfirmation = () => {
    clearTimeout(this.challengePollTimeoutId as TimeoutID);
    if (typeof this.rejectMFAConfirmation === "function") {
      this.rejectMFAConfirmation(new MFAConfirmationCanceledError());
    }
  };

  /**
   * Create a device and return its `deviceId` and `challengeId` for verification
   */
  public createDevice = (
    params: CreateDeviceParams
  ): Promise<CreateDeviceResult> => {
    return this.request(CREATE_DEVICE_PATH, HttpMethod.POST, params);
  };

  /**
   * Verify the device by providing signed OTP received via SMS
   */
  public verifyDevice = (
    deviceId: string,
    params: VerifyDeviceParams
  ): Promise<void> => {
    return this.request(VERIFY_DEVICE_PATH(deviceId), HttpMethod.POST, params);
  };

  /**
   * Create a device challenge and return string to sign by private key
   */
  public createDeviceChallenge = (
    deviceId: string
  ): Promise<DeviceChallenge> => {
    return this.request(
      CREATE_DEVICE_CHALLENGE_PATH(deviceId),
      HttpMethod.POST
    );
  };

  /**
   * Verify the device challenge and update access token
   */
  public verifyDeviceChallenge = async (
    deviceId: string,
    challengeId: string,
    params: VerifyDeviceChallengeParams
  ): Promise<ClientOAuth2.Token> => {
    const { token: accessToken }: VerifyDeviceChallengeResult = await this.request(
      VERIFY_DEVICE_CHALLENGE_PATH(deviceId, challengeId),
      HttpMethod.POST,
      params
    );
    const { refreshToken } = this._token || {};
    return this.setToken(accessToken, refreshToken);
  };

  /**
   * Returns current token used for API requests.
   *
   * @returns  token object which might contain token(s), scope(s), token type and expiration time
   */
  get token(): ClientOAuth2.Token | null {
    return this._token;
  }
}
